### PR TITLE
feat: topics subscription resiliency

### PR DESCRIPTION
**Pull request by anitarua**
Addresses https://github.com/momentohq/client-sdk-swift/issues/106 and part of https://github.com/momentohq/dev-eco-issue-tracker/issues/604 (will need to follow up to update the topics example after this is released).

Creates a Topics GRPC manager class to allow multiple GRPC connections in the pubsub client if needed in the future.
Adds logic for resubscribing on errors that are not cancellation errors.
Adds a way to unsubscribe from a topic (essentially stops the AsyncStream providing the messages).

Manually tested using topics example that it will resubscribe when wifi is turned off for a few minutes and that it will not resubscribe when the user calls unsubscribe to cancel the subscription. 

If you want to test it yourself, the `Package.swift` was modified like so:
```
// swift-tools-version: 5.7
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "momento-topics-example",
    platforms: [.macOS(.v10_15), .iOS(.v13)],
    products: [
        .executable(
            name: "momento-topics-example",
            targets: ["momento-topics-example"]
        ),
    ],
    dependencies: [
        .package(name: "Momento", path: "../.."),
        .package(url: "https://github.com/apple/swift-log.git", from: "1.0.0"),
    ],
    targets: [
        .executableTarget(
            name: "momento-topics-example",
            dependencies: [
                "Momento",
                .product(name: "Logging", package: "swift-log")
            ],
            path: "Sources"
        ),
    ]
)
```

### CODE REVIEW

File: Sources/Momento/messages/responses/ResponseBase.swift [hunk 1]
```diff
@@ -1,4 +1,4 @@
-public class ErrorResponseBase: ErrorResponseBaseProtocol, CustomStringConvertible {
+public class ErrorResponseBase: ErrorResponseBaseProtocol, Error, CustomStringConvertible {
```
Comment (anitarua) [inline Sources/Momento/messages/responses/ResponseBase.swift:1]: this change was to allow Momento errors to be rethrown as described [here](https://github.com/momentohq/dev-eco-issue-tracker/issues/604)

File: Sources/Momento/internal/TopicsGrpcManager.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import NIOHPACK
+
+@available(macOS 10.15, iOS 13, *)
+internal class TopicsGrpcManager {
+    var channel: GRPCChannel
+    var client: CacheClient_Pubsub_PubsubAsyncClient
+    
+    init(
+        credentialProvider: CredentialProviderProtocol,
+        eventLoopGroup: EventLoopGroup,
```
Comment (anitarua) [inline Sources/Momento/internal/TopicsGrpcManager.swift:12]: docs indicate a single event loop group can be shared by multiple Channels: https://swiftinit.org/docs/swift-nio/niocore/eventloop

File: Sources/Momento/internal/PubsubClient.swift [hunk 3]
```diff
@@ -15,57 +15,35 @@ protocol PubsubClientProtocol {
...
-    var configuration: TopicClientConfigurationProtocol
-    var credentialProvider: CredentialProviderProtocol
-    var sharedChannel: GRPCChannel
-    var eventLoopGroup = PlatformSupport.makeEventLoopGroup(loopCount: 1)
-    var client: CacheClient_Pubsub_PubsubAsyncClient
+    let logger = Logger(label: "PubsubClient")
+    let configuration: TopicClientConfigurationProtocol
+    let credentialProvider: CredentialProviderProtocol
+    let eventLoopGroup = PlatformSupport.makeEventLoopGroup(loopCount: 1)
+    let headers = ["agent": "swift:0.1.0"]
```
Comment (pgautier404) [inline Sources/Momento/internal/PubsubClient.swift:31]: Just noticed we haven't updated this.
```suggestion
    let headers = ["agent": "swift:0.3.2"]
```

**Comment by anitarua** [comment]
Just tested against Chris's dev cell and caught an automatic resubscribe when the routing node was redeployed:

```
Subscriber recieved text message: Hello 730
Subscriber recieved text message: Hello 731
Subscriber recieved text message: Hello 732
2024-01-02T14:02:47-0800 debug momento-logger : [Momento] Caught GRPCStatus unavailable (14): Transport became inactive, attempting to resubscribe
2024-01-02T14:02:47-0800 debug momento-logger : [Momento] Successfully resubscribed
2024-01-02T14:02:47-0800 debug momento-logger : [Momento] topic client received a heartbeat
2024-01-02T14:02:47-0800 debug momento-logger : [Momento] topic client received a discontinuity
Subscriber recieved text message: Hello 728
Subscriber recieved text message: Hello 729
Subscriber recieved text message: Hello 730
Subscriber recieved text message: Hello 731
Subscriber recieved text message: Hello 732
Subscriber recieved text message: Hello 733
```

**Review by pgautier404 — APPROVED** [review]
LGTM!

**Pull request merged**